subsys/bluetooth/host/l2cap_br.c

```diff
@@ -704,6 +704,62 @@ static int l2cap_br_conn_req_reply(struct bt_l2cap_chan *chan, uint16_t result)
 	return 0;
 }
 
+#if defined(CONFIG_BT_L2CAP_DYNAMIC_CHANNEL)
+#if defined(CONFIG_BT_L2CAP_LOG_LEVEL_DBG)
+void bt_l2cap_br_chan_set_state_debug(struct bt_l2cap_chan *chan,
+				   bt_l2cap_chan_state_t state,
+				   const char *func, int line)
+{
+	struct bt_l2cap_br_chan *br_chan;
+
+	br_chan = BR_CHAN(chan);
+
+	LOG_DBG("chan %p psm 0x%04x %s -> %s", chan, br_chan->psm,
+		bt_l2cap_chan_state_str(br_chan->state), bt_l2cap_chan_state_str(state));
+
+	/* check transitions validness */
+	switch (state) {
+	case BT_L2CAP_DISCONNECTED:
+		/* regardless of old state always allows this state */
+		break;
+	case BT_L2CAP_CONNECTING:
+		if (br_chan->state != BT_L2CAP_DISCONNECTED) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_L2CAP_CONFIG:
+		if (br_chan->state != BT_L2CAP_CONNECTING) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_L2CAP_CONNECTED:
+		if (br_chan->state != BT_L2CAP_CONFIG &&
+		    br_chan->state != BT_L2CAP_CONNECTING) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	case BT_L2CAP_DISCONNECTING:
+		if (br_chan->state != BT_L2CAP_CONFIG &&
+		    br_chan->state != BT_L2CAP_CONNECTED) {
+			LOG_WRN("%s()%d: invalid transition", func, line);
+		}
+		break;
+	default:
+		LOG_ERR("%s()%d: unknown (%u) state was set", func, line, state);
```

Comment:
I think it's discouraged to manually insert function/line information into logging strings, since the same information can be enabled through the logging subsystem's build configuration. The line numbers aren't even that useful here - what you really want is the before and after states, i.e. I'd suggest to use a "state2str()" conversion helper to log the invalid transitions as human readable strings instead of numbers.

Reply:
Removed parameters line and func.

Reply:
@lylezhu2012 ah, I see now, sorry. I thought the code was logging the line/func of the LOG_*() macro, but this code exists to instead log the place where `set_state()` was called from. You can disregard this review comment then.